File: backend/app/services/filament_estimate.py
"""Estimate missing filament mass/length using material density and filament diameter."""

from __future__ import annotations

import math
from dataclasses import dataclass

DEFAULT_FILAMENT_DIAMETER_MM = 1.75


@dataclass(frozen=True)
class FilamentReconcileResult:
    mass_grams: float | None
    length_mm: float | None
    mass_from_density: bool = False
    length_from_density: bool = False

    @property
    def missing_mass_in_file(self) -> bool:
        return self.mass_grams is None

    @property
    def missing_length_in_file(self) -> bool:
        return self.length_mm is None


def _cross_section_mm2(diameter_mm: float) -> float:
    return math.pi * (diameter_mm / 2.0) ** 2


def mass_g_from_length_mm(
    length_mm: float,
    density_g_cm3: float,
    *,
    diameter_mm: float = DEFAULT_FILAMENT_DIAMETER_MM,
) -> float:
    volume_mm3 = _cross_section_mm2(diameter_mm) * length_mm
    return (volume_mm3 / 1000.0) * density_g_cm3


def length_mm_from_mass_g(
    mass_g: float,
    density_g_cm3: float,
    *,
    diameter_mm: float = DEFAULT_FILAMENT_DIAMETER_MM,
) -> float:
    volume_mm3 = (mass_g / density_g_cm3) * 1000.0
    return volume_mm3 / _cross_section_mm2(diameter_mm)


def reconcile_filament(
    mass_grams: float | None,
    length_mm: float | None,
    density_g_cm3: float | None,
    *,
    diameter_mm: float = DEFAULT_FILAMENT_DIAMETER_MM,
) -> FilamentReconcileResult:
    """Fill missing mass or length when density and the other measurement are known."""

    out_mass = mass_grams
    out_length = length_mm
    mass_from_density = False
    length_from_density = False

    if density_g_cm3 is not None and density_g_cm3 > 0:
        if out_mass is None and out_length is not None and out_length > 0:
            out_mass = mass_g_from_length_mm(out_length, density_g_cm3, diameter_mm=diameter_mm)
            mass_from_density = True
        if out_length is None and out_mass is not None and out_mass > 0:
            out_length = length_mm_from_mass_g(out_mass, density_g_cm3, diameter_mm=diameter_mm)
            length_from_density = True

    return FilamentReconcileResult(
        mass_grams=out_mass,
        length_mm=out_length,
        mass_from_density=mass_from_density,
        length_from_density=length_from_density,
    )

File: frontend/src/lib/filamentEstimate.ts
/** Client-side filament reconciliation (mirrors backend filament_estimate). */

export const DEFAULT_FILAMENT_DIAMETER_MM = 1.75

export type FilamentReconcileResult = {
  massGrams: number | null
  lengthMm: number | null
  massFromDensity: boolean
  lengthFromDensity: boolean
}

export type GcodeFilamentPreview = {
  filament_mass_grams: number | null
  filament_length_mm: number | null
}

function crossSectionMm2(diameterMm: number): number {
  return Math.PI * (diameterMm / 2) ** 2
}

export function massGFromLengthMm(
  lengthMm: number,
  densityGcm3: number,
  diameterMm: number = DEFAULT_FILAMENT_DIAMETER_MM,
): number {
  const volumeMm3 = crossSectionMm2(diameterMm) * lengthMm
  return (volumeMm3 / 1000) * densityGcm3
}

export function lengthMmFromMassG(
  massG: number,
  densityGcm3: number,
  diameterMm: number = DEFAULT_FILAMENT_DIAMETER_MM,
): number {
  const volumeMm3 = (massG / densityGcm3) * 1000
  return volumeMm3 / crossSectionMm2(diameterMm)
}

export function reconcileFilament(
  massGrams: number | null | undefined,
  lengthMm: number | null | undefined,
  densityGcm3: number | null | undefined,
  diameterMm: number = DEFAULT_FILAMENT_DIAMETER_MM,
): FilamentReconcileResult {
  let outMass = massGrams ?? null
  let outLength = lengthMm ?? null
  let massFromDensity = false
  let lengthFromDensity = false

  if (densityGcm3 != null && densityGcm3 > 0) {
    if (outMass == null && outLength != null && outLength > 0) {
      outMass = massGFromLengthMm(outLength, densityGcm3, diameterMm)
      massFromDensity = true
    }
    if (outLength == null && outMass != null && outMass > 0) {
      outLength = lengthMmFromMassG(outMass, densityGcm3, diameterMm)
      lengthFromDensity = true
    }
  }

  return { massGrams: outMass, lengthMm: outLength, massFromDensity, lengthFromDensity }
}

export type FilamentMetadataWarning = {
  id: string
  message: string
  detail?: string
}

/** Warnings for slicer metadata missing from the file (before/after density fill). */
export function filamentMetadataWarnings(
  raw: GcodeFilamentPreview | null,
  reconciled: FilamentReconcileResult,
): FilamentMetadataWarning[] {
  if (!raw) return []
  const warnings: FilamentMetadataWarning[] = []
  if (raw.filament_mass_grams == null && reconciled.massGrams == null) {
    warnings.push({
      id: 'missing-mass',
      message: 'Weight missing',
      detail: 'Not found in G-code — set a default density on the material to estimate',
    })
  } else if (raw.filament_mass_grams == null && reconciled.massFromDensity) {
    warnings.push({
      id: 'mass-from-density',
      message: 'Est. from density',
      detail: 'Weight estimated from length and material density',
    })
  }
  if (raw.filament_length_mm == null && reconciled.lengthMm == null) {
    warnings.push({
      id: 'missing-length',
      message: 'Length missing',
      detail: 'Not found in G-code — set a default density on the material to estimate',
    })
  } else if (raw.filament_length_mm == null && reconciled.lengthFromDensity) {
    warnings.push({
      id: 'length-from-density',
      message: 'Est. from density',
      detail: 'Length estimated from weight and material density',
    })
  }
  return warnings
}

export function densityForMaterialPreset(
  materialPresets: { id: number; default_density_g_cm3?: number | null }[],
  materialPresetId: string,
): number | null {
  if (!materialPresetId) return null
  const id = Number(materialPresetId)
  const preset = materialPresets.find((p) => p.id === id)
  const d = preset?.default_density_g_cm3
  return d != null && d > 0 ? d : null
}
